crates/iceberg/src/spec/schema.rs

```diff
@@ -105,7 +116,16 @@ impl SchemaBuilder {
     }
 
     /// Builds the schema.
-    pub fn build(self) -> Result<Schema> {
+    pub fn build(mut self) -> Result<Schema> {
+        if let Some(start_from) = self.reassign_field_ids_from {
```

Comment:
I agree - I didn't think about the error messages.
It's unfortunately not super straightforward to switch the order, as for example index_by_id needs a full rebuild as both keys and values changes. I think we have two options to achieve the swap:

1. Use `id_reassigner.apply_*` for all other fields that have been build (i.e. `id_to_field`) or
2. Re-build those fields using the same code - i.e. the build option.

I opted for 2.
While it is slightly less performant because we have to visit the structure again, it is much less error prone as we simply re-use existing and well-tested code.

Let me know what you think!

Reply:
@liurenjie1024 I refactored it in my latest commit, however we cannot get rid of the two functions:
`apply_to_aliases` and `apply_to_identifier_fields`. The builder itself is standard now.